File: bfs_dfs_problems/bfs_dfs_problems__maze.cs
// 実行時間 20ms
using static System.Console;
using static System.Linq.Enumerable;
using Tpl = System.Tuple<int, int>;

internal class Program
{
    static void Main()
    {
        int[] hw = ReadLine().Split().Select( int.Parse ).ToArray();
        int[] syx = ReadLine().Split().Select( int.Parse ).ToArray();
        int[] gyx = ReadLine().Split().Select( int.Parse ).ToArray();
        int[] dx = { 0, 0, -1, 1 };
        int[] dy = { -1, 1, 0, 0 };
        int[][] d = new int[hw[0]][];
        char[][] g = new char[hw[0]][];
        var q = new System.Collections.Generic.Queue<Tpl>();

        foreach( int i in Range( 0, hw[0] ) ) {
            g[i] = ReadLine().ToCharArray();
        }
        foreach( int i in Range( 0, hw[0] ) ) {
            d[i] = Repeat( -1, hw[1] ).ToArray();
        }
        d[syx[0]][syx[1]] = 0;
        q.Enqueue( new Tpl( syx[0], syx[1] ) );
        while( 0 < q.Count ) {
            Tpl t = q.Dequeue();

            foreach( int i in Range( 0, 4 ) ) {
                int x = t.Item2 + dx[i];
                int y = t.Item1 + dy[i];

                if( 0 <= x && x < hw[1] && 0 <= y && y < hw[0] && g[y][x] == '.' && d[y][x] < 0 ) {
                    d[y][x] = d[t.Item1][t.Item2] + 1;
                    q.Enqueue( new Tpl( y, x ) );
                }
            }
        }
        // C# 9 以降は $"" 不要
        WriteLine( 0 <= d[gyx[0]][gyx[1]] ? $"{d[gyx[0]][gyx[1]]}" : "No" );
        return;
    }
}

File: bfs_dfs_problems/bfs_dfs_problems__maze.py
# 実行時間 10ms
from collections import deque

h, w = list(map(int, input().split()))
sy, sx = list(map(int, input().split()))
gy, gx = list(map(int, input().split()))
dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]
d = [[-1] * w for _ in [0] * h]
g = [list(input()) for _ in [0] * h]
q = deque()

d[sy][sx] = 0
q.append((sy, sx))
while 0 < len(q):
    t = q.popleft()

    for i in range(4):
        x = t[1] + dx[i]
        y = t[0] + dy[i]

        if 0 <= x < w and 0 <= y < h and g[y][x] == "." and d[y][x] < 0:
            d[y][x] = d[t[0]][t[1]] + 1
            q.append((y, x))

print(d[gy][gx] if 0 <= d[gy][gx] else "No")
